recsyslearn/accuracy/metrics.py

```diff
@@ -0,0 +1,154 @@
+                ),
+                dtype=np.float32)
+
+
+    @classmethod
+    def __ndcg(cls, ranked_list: pd.DataFrame, pos_items: pd.DataFrame, relevance: pd.DataFrame = None, at: int = None) -> float:
+
+        r"""Compute NDCG score, based on https://github.com/recsyspolimi/RecSys_Course_AT_PoliMi implementation.
+        
+
+        Parameters
+        ----------
+        ranked_list : pd.DataFrame
+            Recommendation list of length \"at\".
+
+        pos_items : pd.DataFrame
+            Relevant items per user. Columns: ['user', 'pos_items'].
+
+        relevance: pd.DataFrame
+            Relevance score associated with pos_items, default None.
+
+        at: int
+            Length of reclist to be evaluated, default None.
+
+
+        Return
+        ------
+        The NDCG@at of the single user.
+        """
+
+
+        if relevance is None:
+            relevance = np.ones_like(pos_items, dtype=np.int32)
+        assert len(relevance) == pos_items.shape[0]
+
+        it2rel = {it: r for it, r in zip(pos_items, relevance)}
+
+        rank_scores = np.asarray([it2rel.get(it, 0.0) for it in ranked_list[:at]], dtype=np.float32)
+        ideal_dcg = cls.__dcg(np.sort(relevance)[::-1][:at])
+        rank_dcg = cls.__dcg(rank_scores)
+
+        if rank_dcg == 0:
+            return 0
+
+        ndcg_ = rank_dcg / ideal_dcg
+
+        return ndcg_
+
+
+    @classmethod
+    def evaluate(cls, top_n: pd.DataFrame, pos_items: pd.DataFrame, ats: tuple = (5, 10)) -> pd.Series:
+
+        r"""Compute the NDCG@k of a model by using its recommendation list.
+        Returns the NDCG averaged over users.
+
+
+        Parameters
+        ----------
+        top_n : pd.DataFrame
+            Top N recommendations' lists for every user. Columns: ['user', 'item', 'rank'].
+
+        target_df : pd.DataFrame
+            Target Interaction dataframe of, i.e., items to be recommended. Columns: ['user', 'item'].
+
+        pos_items : pd.DataFrame
+            Relevant items per user. Columns: ['user', 'pos_items'].
+
+        ats: tuple, default (5, 10)
+            The tuple of values at which to evaluate NDCG@k.
+
+
+        Raises
+        ------
+        ColumnsNotExistException
+            If top_n does not contain columns ('user', 'item', 'rank').
+
+        ColumnsNotExistException
+            If target_df does not contain columns ('user', 'item').
+
+        RecListTooShortException
+            If the top_n list does not contain enough items.
+
+
+        Return
+        ------
+        The pd.Series with the NDCG@n averaged over users, in the form ('NDCG@k_0', ..., 'NDCG@k_n')
+        """
+
+        check_columns_exist(top_n, ['user', 'item', 'rank'])
+        check_columns_exist(pos_items, ['user', 'pos_items'])
+
+        min_calculable_at = top_n.groupby('user').size().min()
+        calculable_ats = [
+            k for k in ats 
+            if k <= min_calculable_at   
+        ]
```

Comment:
Compared to the previous version
```
        for k in ats:
            try:
                test_length(top_n, k)
                calculable_ats += [k]
            except ListTooShortException as e:
                print(e)
                continue
        calculable_ats = tuple(calculable_ats)
        ```
        We do not notify the user if one of the ats got removed because it was too long...right? Is that intended to be so?